sacc/sacc.py

```diff
@@ -984,3 +994,114 @@ def add_theta_xi(self, data_type, tracer1, tracer2, theta, x, window=None, trace
+
+    The Covariance will be combined into either a BlockDiagonal covariance or
+    a Diagonal covariance, depending on the inputs.  Either all inputs should
+    have a covariance attached or none of them.
+
+    Parameters
+    ----------
+    *data_sets: Sacc objects
+        The data sets to combined
+
+    labels: List[str]
+        Optional list of strings to append to tracer and metadata names, in case of a clash.
+
+    Returns
+    -------
+    output: Sacc object
+        The combined data set.
+
+    """
+    # Early return of an empty data set object
+    if len(data_sets) == 0:
+        return Sacc()
+
+    data_0 = data_sets[0]
+
+    # Either all the data sets should have covariances or none of them should.  Concatenating
+    # covariances should be straightforward and should always result in a block-diagonal covariance
+    if data_0.has_covariance():
+        if not all(data_set.has_covariance() for data_set in data_sets):
+            raise ValueError("Either all concatenated data sets must have covariances, "
+                             "or none of them")
+    else:
+        if any(data_set.has_covariance() for data_set in data_sets):
+            raise ValueError("Either all concatenated data sets must have covariances, "
+                             "or none of them")
+
+    output = Sacc()
+
+    # Copy the tracers to the new 
+    for i, data_set in enumerate(data_sets):
+        for tracer in data_set.tracers.values():
+
+            # We will be modifying the tracer, so we copy it.
+            tracer = copy.deepcopy(tracer)
+
+            # Optionally add a suffix label to avoid name clashes.
+            if labels is not None:
+                tracer.name += labels[i]
+
+            # Check for duplicate tracer names.
+            # Probably this happens because the user has not provided
+            # any labels to use as tracer suffices.  But it could also
+            # happen if they have chosen really really bad labels
+            if tracer.name in output.tracers:
+                if labels is None:
+                    raise ValueError("There is a name clash between tracers in the data sets."
+                                     "Use the labels option to give new names to them")
+                else:
+                    raise ValueError("After applying your labels there is still a name "
+                                     "clash between tracers in your concatenation.  Try "
+                                     "different labels?")
+            # Build up the combined tracer collection
+            output.add_tracer_object(tracer)
+
+        for d in data_set.data:
+            # Shallow copy because we do not want to clone Window functions,
+            # since they are often shared. The reason we do it at all
+            # is because we may be modifying the tracers names below.
+            # Should we add a label to the data point too?  Or is the tracer
+            # info sufficient?
```

Comment:
We should write a test for this.

Reply:
added something in the existing test, and restructured a little to avoid having to specify underscores manually.